include/geometry/lines_set.h

```diff
@@ -0,0 +1,308 @@
+    }
+
+    void pop_back()
+    {
+        lines_.pop_back();
+    }
+
+    size_t size() const
+    {
+        return lines_.size();
+    }
+
+    bool empty() const
+    {
+        return lines_.empty();
+    }
+
+    void reserve(size_t size)
+    {
+        lines_.reserve(size);
+    }
+
+    void resize(size_t size)
+    {
+        lines_.resize(size);
+    }
+
+    void clear()
+    {
+        lines_.clear();
+    }
+
+    template<typename... Args>
+    void emplace_back(Args&&... args)
+    {
+        lines_.emplace_back(args...);
+    }
+
+    std::vector<LineType>::iterator erase(std::vector<LineType>::const_iterator first, std::vector<LineType>::const_iterator last)
+    {
+        return lines_.erase(first, last);
+    }
+
+    LinesSet& operator=(const LinesSet& other)
+    {
+        lines_ = other.lines_;
+        return *this;
+    }
+
+    LinesSet& operator=(LinesSet&& other)
+    {
+        lines_ = std::move(other.lines_);
+        return *this;
+    }
+
+    LineType& operator[](size_t index)
+    {
+        return lines_[index];
+    }
+
+    const LineType& operator[](size_t index) const
+    {
+        return lines_[index];
+    }
+
+    LineType& newLine()
+    {
+        lines_.emplace_back();
+        return lines_.back();
+    }
+
+    /*! \brief Return the amount of points in all lines */
+    size_t pointCount() const;
+
+    /*!
+     * Remove a line from the list and move the last line to its place
+     * \warning changes the order of the lines!
+     */
+    void removeAt(size_t index);
+
+    /*! \brief Add a simple line consisting of two points */
+    void addSegment(const Point2LL& from, const Point2LL& to);
+
+    /*! \brief Get the total length of all the lines */
+    coord_t length() const;
+
+    void splitIntoSegments(LinesSet<OpenPolyline>& result) const;
+    LinesSet<OpenPolyline> splitIntoSegments() const;
+
+    /*! \brief Removes overlapping consecutive line segments which don't delimit a positive area */
+    void removeDegenerateVerts();
+
+    Shape offset(coord_t distance, ClipperLib::JoinType join_type = ClipperLib::jtMiter, double miter_limit = 1.2) const;
+
+    /*!
+     * Utility method for creating the tube (or 'donut') of a shape.
+     * \param inner_offset Offset relative to the original shape-outline towards the inside of the shape. Sort-of like a negative normal offset, except it's the offset part that's
+     * kept, not the shape. \param outer_offset Offset relative to the original shape-outline towards the outside of the shape. Comparable to normal offset. \return The resulting
+     * polygons.
+     */
+    Shape tubeShape(const coord_t inner_offset, const coord_t outer_offset) const;
+
+    void translate(const Point2LL& delta);
+
+    /*!
+     * \brief Utility method to add all the lines to a ClipperLib::Clipper object
+     * \note This method needs to be public but you shouldn't need to use it from outside
```

Comment:
I had to do that in order to make this call possible:
https://github.com/Ultimaker/CuraEngine/blob/97eb08f0cb2a01cadb6ebc1e01078acf4b127f6c/src/geometry/lines_set.cpp#L185C9-L185C56
Otherwise you can't call it on an other instance, even if you are in the same class. So adding a friend wouldn't help in this case because it would friend itself.
But I'm quite sure I missed something, because sometimes I know you can call private members of other instances...

Reply:
Could not find a clean way to overcome this, that would actually add a lot of dirty code.. Rejecting.